File: crates/lilavati-core/src/panchang.rs
//! Panchang computation engine — the 5 elements of the Hindu calendar.
//!
//! All computations use sunrise JD as the reference point,
//! because the Hindu day starts at sunrise.

use crate::angles::normalize;
use crate::constants::*;
use crate::ephemeris::{self, Planet};
use crate::search::{find_crossing_backward, find_crossing_forward};
use crate::types::*;

/// Span of each element in degrees.
const TITHI_SPAN: f64 = 12.0; // 360 / 30
const NAKSHATRA_SPAN: f64 = 360.0 / 27.0; // 13°20'
const YOGA_SPAN: f64 = 360.0 / 27.0;
const KARANA_SPAN: f64 = 6.0; // 360 / 60

/// Max days to search for transition times.
const MAX_SEARCH_DAYS: f64 = 2.0;

/// Compute complete Panchang at a given sunrise JD.
pub fn compute(sunrise_jd: f64, weekday: u32) -> PanchangResult {
    let vara = compute_vara(weekday);
    let tithi = compute_tithi(sunrise_jd);
    let nakshatra = compute_nakshatra(sunrise_jd);
    let yoga = compute_yoga(sunrise_jd);
    let karana = compute_karana(sunrise_jd);

    PanchangResult {
        sun: SunData {
            sunrise_jd,
            sunset_jd: 0.0,          // filled by caller
            day_duration_hours: 0.0, // filled by caller
        },
        vara,
        tithi,
        nakshatra,
        yoga,
        karana,
    }
}

// --- Vara (weekday) ---

fn compute_vara(weekday: u32) -> VaraInfo {
    let idx = weekday as usize % 7;
    VaraInfo {
        number: weekday,
        name: VARA_NAMES[idx],
        english: VARA_ENGLISH[idx],
    }
}

// --- Tithi ---

/// Get the tithi angle (Moon - Sun tropical longitude) at a given JD.
fn tithi_angle(jd: f64) -> f64 {
    let moon = ephemeris::tropical_longitude(jd, Planet::Moon);
    let sun = ephemeris::tropical_longitude(jd, Planet::Sun);
    normalize(moon - sun)
}

fn compute_tithi(jd: f64) -> TithiInfo {
    let angle = tithi_angle(jd);
    let tithi_idx = (angle / TITHI_SPAN) as u32; // 0-indexed (0-29)
    let tithi_num = tithi_idx + 1; // 1-indexed (1-30)

    let paksha = if tithi_num <= 15 { "Shukla" } else { "Krishna" };
    let tithi_name = TITHI_NAMES[tithi_idx as usize];
    let full_name = format!("{} {}", paksha, tithi_name);

    // Transition times
    let target_start = tithi_idx as f64 * TITHI_SPAN;
    let target_end = (tithi_idx + 1) as f64 * TITHI_SPAN;

    let angle_fn = |jd: f64| -> f64 { tithi_angle(jd) };

    let start_jd =
        find_crossing_backward(jd, target_start, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd - 1.0);
    let end_jd =
        find_crossing_forward(jd, target_end, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd + 1.0);

    TithiInfo {
        number: tithi_num,
        name: full_name,
        paksha,
        start_jd,
        end_jd,
    }
}

// --- Nakshatra ---

/// Get Moon's sidereal longitude for Nakshatra computation.
fn nakshatra_angle(jd: f64) -> f64 {
    ephemeris::sidereal_longitude(jd, Planet::Moon)
}

fn compute_nakshatra(jd: f64) -> NakshatraInfo {
    let moon_sid = nakshatra_angle(jd);
    let nak_idx = (moon_sid / NAKSHATRA_SPAN) as u32; // 0-indexed (0-26)
    let nak_num = nak_idx + 1;

    // Pada (quarter)
    let pada_span = NAKSHATRA_SPAN / 4.0;
    let offset_in_nak = moon_sid - (nak_idx as f64 * NAKSHATRA_SPAN);
    let pada = ((offset_in_nak / pada_span) as u32 + 1).min(4);

    // Transition times
    let target_start = nak_idx as f64 * NAKSHATRA_SPAN;
    let mut target_end = (nak_idx + 1) as f64 * NAKSHATRA_SPAN;
    if target_end >= 360.0 {
        target_end -= 360.0;
    }

    let angle_fn = |jd: f64| -> f64 { nakshatra_angle(jd) };

    let start_jd =
        find_crossing_backward(jd, target_start, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd - 1.0);
    let end_jd =
        find_crossing_forward(jd, target_end, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd + 1.0);

    NakshatraInfo {
        number: nak_num,
        name: NAKSHATRA_NAMES[nak_idx as usize],
        pada,
        lord: NAKSHATRA_LORDS[nak_idx as usize],
        start_jd,
        end_jd,
    }
}

// --- Yoga ---

/// Yoga angle: (Sun sidereal + Moon sidereal) mod 360.
fn yoga_angle(jd: f64) -> f64 {
    let sun_sid = ephemeris::sidereal_longitude(jd, Planet::Sun);
    let moon_sid = ephemeris::sidereal_longitude(jd, Planet::Moon);
    normalize(sun_sid + moon_sid)
}

fn compute_yoga(jd: f64) -> YogaInfo {
    let angle = yoga_angle(jd);
    let yoga_idx = (angle / YOGA_SPAN) as u32; // 0-indexed (0-26)
    let yoga_num = yoga_idx + 1;

    let target_start = yoga_idx as f64 * YOGA_SPAN;
    let mut target_end = (yoga_idx + 1) as f64 * YOGA_SPAN;
    if target_end >= 360.0 {
        target_end -= 360.0;
    }

    let angle_fn = |jd: f64| -> f64 { yoga_angle(jd) };

    let start_jd =
        find_crossing_backward(jd, target_start, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd - 1.0);
    let end_jd =
        find_crossing_forward(jd, target_end, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd + 1.0);

    YogaInfo {
        number: yoga_num,
        name: YOGA_NAMES[yoga_idx as usize],
        start_jd,
        end_jd,
    }
}

// --- Karana ---

fn compute_karana(jd: f64) -> KaranaInfo {
    let angle = tithi_angle(jd);
    let karana_60 = (angle / KARANA_SPAN) as u32; // 0-indexed (0-59)

    let karana_name = karana_name_from_number(karana_60);
    let karana_number = KARANA_NAMES
        .iter()
        .position(|&n| n == karana_name)
        .map(|i| i as u32 + 1)
        .unwrap_or(1);

    // Transition times
    let target_start = karana_60 as f64 * KARANA_SPAN;
    let mut target_end = (karana_60 + 1) as f64 * KARANA_SPAN;
    if target_end >= 360.0 {
        target_end -= 360.0;
    }

    let angle_fn = |jd: f64| -> f64 { tithi_angle(jd) };

    let start_jd =
        find_crossing_backward(jd, target_start, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd - 1.0);
    let end_jd =
        find_crossing_forward(jd, target_end, &angle_fn, MAX_SEARCH_DAYS).unwrap_or(jd + 1.0);

    KaranaInfo {
        number: karana_number,
        name: karana_name,
        start_jd,
        end_jd,
    }
}

/// Map a karana number (0-59) to its name.
///
/// Karana cycle in a lunar month:
///   0: Kimstughna (fixed — first half of Shukla Pratipada)
///   1-56: 7 rotating karanas repeated 8 times
///   57: Shakuni (fixed)
///   58: Chatushpada (fixed)
///   59: Nagava (fixed)
fn karana_name_from_number(karana_60: u32) -> &'static str {
    match karana_60 {
        0 => "Kimstughna",
        1..=56 => {
            let rotating_idx = ((karana_60 - 1) % 7) as usize;
            KARANA_NAMES[rotating_idx]
        }
        57 => "Shakuni",
        58 => "Chatushpada",
        59 => "Nagava",
        _ => "Bava", // Should never happen
    }
}

File: python/panchang/core/panchang.py
"""Panchang computation engine for Panchang.

Computes the 5 elements of the Hindu Panchang:
  1. Vara (weekday)
  2. Tithi (lunar day) — based on Sun-Moon angular separation
  3. Nakshatra (lunar mansion) — based on Moon's sidereal longitude
  4. Yoga (Sun-Moon combination) — based on sum of sidereal longitudes
  5. Karana (half-tithi)

All computation is delegated to the Rust _core extension.
"""

from __future__ import annotations

from datetime import datetime
from zoneinfo import ZoneInfo

from panchang._core import py_compute_panchang, py_init
from panchang.core.ephemeris import EphemerisEngine, jd_to_datetime
from panchang.core.sun import _tz_offset_for_date, compute_sun_data
from panchang.types import (
    KaranaInfo,
    Location,
    NakshatraInfo,
    Paksha,
    PanchangData,
    TithiInfo,
    VaraInfo,
    YogaInfo,
)


def compute_panchang(
    dt: datetime,
    location: Location,
    engine: EphemerisEngine | None = None,
) -> PanchangData:
    """Compute complete Panchang for a date and location.

    Args:
        dt: Date/datetime for computation.
        location: Geographic location.
        engine: Optional ephemeris engine (uses default Lahiri if not provided).

    Returns:
        PanchangData with all 5 elements plus sunrise/sunset.
    """
    py_init(None)
    tz = ZoneInfo(location.tz)

    if dt.tzinfo is None:
        local_dt = dt.replace(tzinfo=tz)
    else:
        local_dt = dt.astimezone(tz)

    # Weekday: Python Monday=0, we need Sunday=0
    weekday = (local_dt.weekday() + 1) % 7

    utc_offset = _tz_offset_for_date(location.tz, local_dt.year, local_dt.month, local_dt.day)

    raw = py_compute_panchang(
        local_dt.year,
        local_dt.month,
        local_dt.day,
        location.lat,
        location.lng,
        location.altitude,
        utc_offset,
        weekday,
    )

    # Convert raw dict to typed models
    sun_data = compute_sun_data(local_dt, location, engine)

    vara_raw = raw["vara"]
    vara = VaraInfo(
        number=vara_raw["number"],
        name=vara_raw["name"],
        english=vara_raw["english"],
    )

    tithi_raw = raw["tithi"]
    tithi_num = tithi_raw["number"]
    paksha = Paksha.SHUKLA if tithi_num <= 15 else Paksha.KRISHNA
    tithi = TithiInfo(
        number=tithi_num,
        name=tithi_raw["name"],
        paksha=paksha,
        start=jd_to_datetime(tithi_raw["start_jd"]).astimezone(tz),
        end=jd_to_datetime(tithi_raw["end_jd"]).astimezone(tz),
    )

    nak_raw = raw["nakshatra"]
    nakshatra = NakshatraInfo(
        number=nak_raw["number"],
        name=nak_raw["name"],
        pada=nak_raw["pada"],
        lord=nak_raw["lord"],
        start=jd_to_datetime(nak_raw["start_jd"]).astimezone(tz),
        end=jd_to_datetime(nak_raw["end_jd"]).astimezone(tz),
    )

    yoga_raw = raw["yoga"]
    yoga = YogaInfo(
        number=yoga_raw["number"],
        name=yoga_raw["name"],
        start=jd_to_datetime(yoga_raw["start_jd"]).astimezone(tz),
        end=jd_to_datetime(yoga_raw["end_jd"]).astimezone(tz),
    )

    karana_raw = raw["karana"]
    karana = KaranaInfo(
        number=karana_raw["number"],
        name=karana_raw["name"],
        start=jd_to_datetime(karana_raw["start_jd"]).astimezone(tz),
        end=jd_to_datetime(karana_raw["end_jd"]).astimezone(tz),
    )

    return PanchangData(
        date=local_dt.strftime("%Y-%m-%d"),
        location=location,
        sun=sun_data,
        vara=vara,
        tithi=tithi,
        nakshatra=nakshatra,
        yoga=yoga,
        karana=karana,
    )
